pkg/net/cidrs/triemap.go

```diff
@@ -0,0 +1,262 @@
+	// see: https://planetscale.com/blog/generics-can-make-your-go-code-slower
+	// The maps below map from int in this trie to generic value type V
+	//
+	// This is also cheaper in many cases because int will be smaller than V
+	// so we can store V only once in the map here, and int indexes into those
+	// maps in the trie structure, given than many trie nodes will map to the same
+	// V, as our target use-case is CIDR-to-cloud-region
+	trieMap trieMap
+
+	// simple inline bimap of int keys to V values
+	//
+	// the inner trie stores an int key index into keyToValue
+	//
+	// valueToKey is to cheapen checking if we've already inserted a given V
+	// and use the same key
+	keyToValue map[int]V
+	valueToKey map[V]int
+}
+
+// NewTrieMap[V] returns a new, properly allocated TrieMap[V]
+func NewTrieMap[V comparable]() *TrieMap[V] {
+	return &TrieMap[V]{
+		keyToValue: make(map[int]V),
+		valueToKey: make(map[V]int),
+	}
+}
+
+// Insert inserts value into TrieMap by index cidr
+// You can later match a netip.Addr to value with GetIP
+func (t *TrieMap[V]) Insert(cidr netip.Prefix, value V) {
+	key, alreadyHave := t.valueToKey[value]
+	if !alreadyHave {
+		// next key = length of map
+		// this structure is insert-only
+		key = len(t.keyToValue)
+		t.valueToKey[value] = key
+		t.keyToValue[key] = value
+	}
+	t.trieMap.Insert(cidr, key)
+}
+
+// GetIP returns the associated value for the matching cidr if any with contains=true,
+// or else the default value of V and contains=false
+func (t *TrieMap[V]) GetIP(ip netip.Addr) (value V, contains bool) {
+	// NOTE: this is written so as not to shadow contains locally
+	// and so we can use value as a default-value for V without
+	// another variable, using the name also to document the return
+	key, c := t.trieMap.GetIP(ip)
+	contains = c
+	if !contains {
+		return
+	}
+	value = t.keyToValue[key]
+	return
+}
+
+// trieMap is the core implementation, but it only stores netip.Prefix : int
+type trieMap struct {
+	// surely ipv4 and ipv6 will be enough in our lifetime?
+	ipv4Root *trieNode
+	ipv6Root *trieNode
+}
+
+// TODO: path compression
+type trieNode struct {
+	// children for 0 and 1 bits
+	child0 *trieNode
+	child1 *trieNode
+	// both of these values will be set together or not set
+	// so we place them in a sub struct to save memory at the cost
+	// of chasing one additional pointer per trie node checked
+	value *nodeValue
+}
+
+type nodeValue struct {
+	cidr netip.Prefix
+	key  int
+}
+
+func (t *trieMap) Insert(cidr netip.Prefix, key int) {
+	if cidr.Addr().Is4() {
+		t.insertIPV4(cidr, key)
+	} else {
+		t.insertIPV6(cidr, key)
+	}
+}
+
+func (t *trieMap) insertIPV4(cidr netip.Prefix, key int) {
+	// ensure root node
+	if t.ipv4Root == nil {
+		t.ipv4Root = &trieNode{}
+	}
+
+	// walk bits high to low, inserting matching ip path up to mask bits
+	curr := t.ipv4Root
+	ip := cidr.Addr().As4()
+	ipInt := uint32(ip[3]) | uint32(ip[2])<<8 | uint32(ip[1])<<16 | uint32(ip[0])<<24
```

Comment:
constructing the ipInt from the bytes could use a comment.

Reply:
in the ipv6 case (lines 163/164) we are traversing the 16 bytes but here seem to create an integer out of the 4 bytes and THEN traverse it... do we want to avoid that with code similar to the ipv6 case? (curious)

Reply:
thanks Ben!

Reply:
I'm splitting the IPv4 and IPv6 benchmarks so we can more clearly measure these changes.

After splitting, doing IPv4 the byte by byte way is ~ 55 ns/op vs ~ 39 ns/op pre-casting to uint32 (some variance in bench runs but consistently faster pre-casting).

IPv6 is >200 ns/op vs ~170 ns / op pre-casting to two uint64 (again variance, but faster pre-casting).

All of these values are more than excellent for our purposes, but IMHO it's more idiomatic for this type of algorithm to mask fixed width integers than byte by byte as well.